File: Data Structures & Algorithms/balanced-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

final class TreeInfo {
    boolean balanced;
    int height;

    public TreeInfo(boolean balanced, int height) {
        this.balanced = balanced;
        this.height = height;
    }
}

class Solution {
    public TreeInfo dfs(TreeNode node) {
            if (node == null) {
                return new TreeInfo(true, 0);
            }
            TreeInfo left = dfs(node.left);
            TreeInfo right = dfs(node.right);
            boolean balanced = (left.balanced && right.balanced) 
                && Math.abs(left.height - right.height) <= 1;
            return new TreeInfo(balanced, 1 + Math.max(left.height, right.height));


    }
    public boolean isBalanced(TreeNode root) {

        return dfs(root).balanced;
    }
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
import math 
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        
        def dfs(node, rng):
            if not node:
                return True
            if not rng[0] < node.val < rng[1]:
                return False
            leftRng = [-math.inf, node.val]
            rightRng = [node.val, math.inf]
            return dfs(node.left, [rng[0], node.val]) and dfs(node.right, [node.val, rng[1]])
        
        return dfs(root, [-math.inf, math.inf])